engine/lib/users.php

```diff
@@ -445,14 +445,28 @@ function user_create_hook_add_site_relationship($event, $object_type, $object) {
  * @access private
  */
 function user_avatar_hook($hook, $entity_type, $returnvalue, $params) {
-	$user = $params['entity'];
-	$size = $params['size'];
+	$user = elgg_extract('entity', $params);
+	$size = elgg_extract('size', $params, 'medium');
 
-	if (isset($user->icontime)) {
+	if (!$user instanceof ElggUser) {
+		return;
+	}
+
+	$default_url = elgg_get_simplecache_url("icons/user/default{$size}.gif");
+	if (!isset($user->icontime)) {
+		return $default_url;
+	}
+
+	if (_elgg_view_may_be_altered('resources/avatar/view', 'resources/avatar/view.php')) {
 		return "avatar/view/$user->username/$size/$user->icontime";
```

Comment:
Will add. FYI, tested manually by adding a resource view in a plugin. Didn't test with hooks.